tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,170 @@
+    
+    Parameters:
+        events (ndarray): ndarray of shape [num_events, num_event_channels].
+        sensor_size (tuple): a 3-tuple of x,y,p for sensor_size.
+        target_size (tuple): a 2-tuple of x,y denoting new down-sampled size for events to be
+                             re-scaled to (new_width, new_height).
+        dt (float): step size for simulation, in ms.
+        differentiator_time_bins (int): number of equally spaced time bins with respect to the dt 
+                                        to be used for the differentiator.
+        noise_threshold (int): number of events before a spike representing a new event is emitted.
+        
+    Returns:
+        the downsampled input events using the differentiator method.
+    """
+        
+    assert "x" and "y" and "t" in events.dtype.names
+    assert np.logical_and(np.remainder(differentiator_time_bins, 1) == 0, differentiator_time_bins >= 1)
+    
+    events = events.copy()
+    
+    # Call integrator method
+    dt_scaling, events_integrated = integrator_downsample(events, sensor_size=sensor_size, target_size=target_size, 
+                                                          dt=(dt / differentiator_time_bins), 
+                                                          noise_threshold=noise_threshold, differentiator_call=True)
+    
+    if dt_scaling:
+        dt *= 1000
+        
+    num_frames = int(events_integrated[-1][0] // dt + 1)
+    frame_histogram = np.zeros((num_frames, *np.flip(target_size), 2))
+        
+    for event in events_integrated:
+        differentiated_time, event_histogram = event
+        time = int(differentiated_time // dt)
+        
+        # Separate events based on polarity and apply Heaviside
+        event_hist_pos = (np.maximum(event_histogram, 0)).clip(max=1)
+        event_hist_neg = (-np.minimum(event_histogram, 0)).clip(max=1)
+        
+        frame_histogram[time,...,1] += event_hist_pos
+        frame_histogram[time,...,0] += event_hist_neg
+        
+    # Differences between subsequent frames
+    frame_differences = np.diff(frame_histogram, axis=0).clip(min=0)
+    
+    # Restructuring numpy array to structured array
+    time_index, y_new, x_new, polarity_new = np.nonzero(frame_differences)
+    
+    events_new = np.column_stack((x_new, y_new, polarity_new.astype(dtype=bool), time_index * dt))
+    
+    return unstructured_to_structured(events_new.copy(), dtype=[("x", "<i4"), ("y", "<i4"), ("p", "<i4"), ("t", "<i4")])
+    
+def integrator_downsample(events: np.ndarray, sensor_size: tuple, target_size: tuple, dt: float, noise_threshold: int = 0, 
+                          differentiator_call: bool = False):
+    """Downsample using an integrate-and-fire (I-F) neuron model with a noise threshold similar to 
+    the membrane potential threshold in the I-F model. Multiply x/y values by a spatial_factor 
+    obtained by dividing sensor size by the target size.
```

Comment:
Can you explain what's happening in a different way? Is the spatial pooling happening before or after the integration? What does it mean that the noise threshold is similar to the membrane potential threshold? I understand filtering events using an IF neuron like in the [Decimation](https://tonic.readthedocs.io/en/latest/generated/tonic.transforms.Decimation.html#tonic.transforms.Decimation) transform, where essentially every xth event causes a spike. What is the purpose of the noise threshold? It would be great to understand the idea of this transform without having to read your paper

Reply:
So, from neuroscience, an IF neuron fires only when the membrane potential is above a certain threshold. Similarly, if the number of events in a pixel in the running event buffer (as I explained in the other comment) exceeds a certain noise threshold, a spike is emitted that I consider to be the actual event point. This is relevant if there is spatio-temporal downsampling, otherwise the method resolves to the `Downsampling` transform. Although I understand that it would be great to understand without having to read my paper, the paper contains figures to elucidate these concepts.

Reply:
I will provide a more comprehensive explanation in the docstrings in the next commit. Please let me know your thoughts as well regarding which points need more explanation.